bin/deployables

```diff
@@ -83,600 +99,829 @@ region=${DEPLOY_AWS_REGION}
+    eval keyfile="~/.ssh/id_${fingerprint}"
 
-    done
+    if [[ ! -f ${keyfile} ]] ; then
+      echo "Error: Unable to find machine user key file"
+      echo "- fingerprint: ${fingerprint}"
+      echo "- keyfile: ${keyfile}"
+
+      eval keyfile="~/.ssh/id_circleci_github"
+
+      if [[ ! -f ${keyfile} ]] ; then
+        echo "Error: Unable to find circle github key file"
+        echo "- keyfile: ${keyfile}"
+        return 1
+      fi
+
+    fi
+
+    echo "Using GITHUB_MACHINE_USER_KEY=${keyfile}"
+
+    docker build \
+      --rm=false \
+      --no-cache \
+      --build-arg GITHUB_MACHINE_USER_KEY="$(< ${keyfile} )" \
+      --build-arg NPM_TOKEN=$NPM_TOKEN \
+      -t ${DEPLOY_DOCKER_LOCAL_TAG} \
+      -f ${DEPLOY_DOCKERFILE} \
+      .
+
+  else
+
+    docker build \
+      --rm=false \
+      --build-arg NPM_TOKEN=$NPM_TOKEN \
+      -t ${DEPLOY_DOCKER_LOCAL_TAG} \
+      -f ${DEPLOY_DOCKERFILE} \
+      .
+
+  fi
 
-    echo "Error: Service update took too long."
-    exit 1
 }
 
 
-##
-## public functions
-##
+function _deploy_image () {
+  #
+  # push images
+  #
 
-function docker_build {
+  if [ ! -z "${DEPLOY_RETAG_AND_PUSH}" ] ; then
+
+    remote_tag="${DEPLOY_ECR_HOST}/${DEPLOY_APP_NAME}:${DEPLOY_RETAG_REMOTE_TAG}"
+    target_tag="${DEPLOY_ECR_HOST}/${DEPLOY_APP_NAME}:${DEPLOY_RETAG_TARGET_TAG}"
 
     echo
-    echo Building docker image
-	echo DEPLOY_DOCKERFILE: ${DEPLOY_DOCKERFILE}
-	echo DEPLOY_DOCKER_LOCAL_TAG: ${DEPLOY_DOCKER_LOCAL_TAG}
+    echo DEPLOY_RETAG_AND_PUSH: ${DEPLOY_RETAG_AND_PUSH}
+    echo remote_tag: ${remote_tag}
+    echo target_tag: ${target_tag}
     echo
 
-	if [ -z "${DEPLOY_DOCKERFILE}" ] ; then
-		echo "Error: DEPLOY_DOCKERFILE is required"
-		return 1
-	fi
+    ecr_pull_image \
+      $remote_tag
 
-	if [ -z "${DEPLOY_DOCKER_LOCAL_TAG}" ] ; then
-		echo "Error: DEPLOY_DOCKER_LOCAL_TAG is required"
-		return 1
-	fi
+    ecr_tag_image \
+      $remote_tag \
+      $target_tag
 
-	if [[ ! -z "${DEPLOY_GITHUB_MACHINE_USER_KEY_FINGERPRINT}" ]] ; then
+    ecr_push_image \
+      $target_tag
 
-		fingerprint="$( echo ${DEPLOY_GITHUB_MACHINE_USER_KEY_FINGERPRINT} | tr -d : )"
 
-		eval keyfile="~/.ssh/id_${fingerprint}"
+  else
+    target_tag="${DEPLOY_ECR_HOST}/${DEPLOY_APP_NAME}:$(_get_target_image_tag)"
 
-		if [[ ! -f ${keyfile} ]] ; then
-			echo "Error: Unable to find machine user key file"
-			echo "- fingerprint: ${fingerprint}"
-			echo "- keyfile: ${keyfile}"
+    echo
+    echo Pushing primary tag
+    echo target_tag: ${target_tag}
+    echo
 
-            eval keyfile="~/.ssh/id_circleci_github"
+    ecr_tag_image \
+      $DEPLOY_DOCKER_LOCAL_TAG \
+      $target_tag
 
-            if [[ ! -f ${keyfile} ]] ; then
-                echo "Error: Unable to find circle github key file"
-                echo "- keyfile: ${keyfile}"
-                return 1
-            fi
+    ecr_push_image \
+      $target_tag
 
-		fi
+    if [ ! -z "$DEPLOY_PUSH_SECONDARY_TAG" ] ; then
```

Comment:
Is there not a case for `DEPLOY_RETAG_AND_PUSH` and `DEPLOY_PUSH_SECONDARY_TAG` to both be set?